### PR TITLE
Add Set_cert key_pair_id parameter to sync spdm 1.3

### PR DESCRIPTION
**Pull request by Wenxing-hou**
Fix: #2292

### CODE REVIEW

**Comment by jyao1** [comment]
Is this a full patch?

**Comment by Wenxing-hou** [comment]
> Is this a full patch?

Thanks. I have updated the patch to make it completely.

File: library/spdm_responder_lib/libspdm_rsp_set_certificate.c [hunk 1]
```diff
@@ -151,6 +152,19 @@ libspdm_return_t libspdm_get_response_set_certificate(libspdm_context_t *spdm_co
     root_cert_hash_size = libspdm_get_hash_size(
         spdm_context->connection_info.algorithm.base_hash_algo);
 
+    if ((libspdm_get_connection_version(spdm_context) >= SPDM_MESSAGE_VERSION_13) &&
+        (spdm_context->connection_info.multi_key_conn_rsp)) {
+        key_pair_id = spdm_request->header.param2;
+        /*write KeyPairID.*/
+        spdm_context->local_context.local_key_pair_id[slot_id] = key_pair_id;
+    } else {
+        if (spdm_request->header.param2 != 0) {
```
Comment (jyao1) [inline library/spdm_responder_lib/libspdm_rsp_set_certificate.c:161]: this check should only in 1.3+.

Comment (Wenxing-hou) [reply library/spdm_responder_lib/libspdm_rsp_set_certificate.c]: Thanks. I have deleted the check in 1.2.

File: library/spdm_responder_lib/libspdm_rsp_set_certificate.c [hunk 2]
```diff
@@ -151,6 +152,13 @@ libspdm_return_t libspdm_get_response_set_certificate(libspdm_context_t *spdm_co
     root_cert_hash_size = libspdm_get_hash_size(
         spdm_context->connection_info.algorithm.base_hash_algo);
 
+    if ((libspdm_get_connection_version(spdm_context) >= SPDM_MESSAGE_VERSION_13) &&
+        (spdm_context->connection_info.multi_key_conn_rsp)) {
+        key_pair_id = spdm_request->header.param2;
+        /*write KeyPairID.*/
+        spdm_context->local_context.local_key_pair_id[slot_id] = key_pair_id;
+    }
```
Comment (jyao1) [inline library/spdm_responder_lib/libspdm_rsp_set_certificate.c:160]: I think we need to check version >= 1.3 && multi_key_conn_rsp == false.

File: library/spdm_responder_lib/libspdm_rsp_set_certificate.c [hunk 3]
```diff
@@ -151,6 +152,13 @@ libspdm_return_t libspdm_get_response_set_certificate(libspdm_context_t *spdm_co
     root_cert_hash_size = libspdm_get_hash_size(
         spdm_context->connection_info.algorithm.base_hash_algo);
 
+    if ((libspdm_get_connection_version(spdm_context) >= SPDM_MESSAGE_VERSION_13) &&
+        (spdm_context->connection_info.multi_key_conn_rsp)) {
+        key_pair_id = spdm_request->header.param2;
+        /*write KeyPairID.*/
+        spdm_context->local_context.local_key_pair_id[slot_id] = key_pair_id;
```
Comment (jyao1) [inline library/spdm_responder_lib/libspdm_rsp_set_certificate.c:159]: I suggest to add comment: key_pair_id / slot_id check will be done in the future.

File: library/spdm_responder_lib/libspdm_rsp_set_certificate.c (hunk 2 above)
Comment (Wenxing-hou) [reply library/spdm_responder_lib/libspdm_rsp_set_certificate.c:167]: Thanks. I have added the check.

File: library/spdm_responder_lib/libspdm_rsp_set_certificate.c (hunk 3 above)
Comment (Wenxing-hou) [reply library/spdm_responder_lib/libspdm_rsp_set_certificate.c]: Thanks. I have added the comment.

File: library/spdm_responder_lib/libspdm_rsp_set_certificate.c [hunk 4]
```diff
@@ -151,6 +152,20 @@ libspdm_return_t libspdm_get_response_set_certificate(libspdm_context_t *spdm_co
     root_cert_hash_size = libspdm_get_hash_size(
         spdm_context->connection_info.algorithm.base_hash_algo);
 
+    if (libspdm_get_connection_version(spdm_context) >= SPDM_MESSAGE_VERSION_13) {
+        key_pair_id = spdm_request->header.param2;
+        if (spdm_context->connection_info.multi_key_conn_rsp) {
+            /*write KeyPairID. And key_pair_id/slot_id check will be done in the future.*/
+            spdm_context->local_context.local_key_pair_id[slot_id] = key_pair_id;
```
[the remark this answers is not in the document]
Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_set_certificate.c:159]: Agree. It should just check if key_pair_id is 0.

[the remark this answers is not in the document]
Comment (Wenxing-hou) [reply library/spdm_responder_lib/libspdm_rsp_set_certificate.c]: Thanks. I have updated the code.

**Review by steven-bellock — APPROVED** [review]

**Comment by jyao1** [comment]
Have you validated spdm-emu?

**Pull request merged**